src/midiToNotes.js

```diff
@@ -169,37 +178,120 @@ const MidiToNotes = (function () {
 
         if (currentNote) {
           currentNote.endPitch = pitch;
+          currentNote.endPitchBend = pitchBend;
         } else {
           currentNote = {
             startTime: event.time,
             startPitch: pitch,
             endPitch: pitch,
+            startPitchBend: pitchBend,
+            endPitchBend: pitchBend
           };
         }
       }
     }
   }
 
+  /***
+   * Collects all pitch bend events in the MIDI and converts the 
+   * pitch bend value into semitones.
+   */
+  function collectPitchBendEvents(sortedMidiEvents) {
+    MidiToNotes.pitchBendEvents = [];
+
+    for (const event of sortedMidiEvents) {
+      if (getEventType(event) === "pitchBend") {
+        //A MIDI pitch bend event consists of two bytes 0aaaaaaa and 0bbbbbbb.
+        //These are combined (bbbbbbbaaaaaaa) and converted to semitones
+        //according to the range specified in the settings.
+        const midiPitchBend = ((event.data[1] << 7 | event.data[0]) - 8192) / 8192.0;
+        const pitchEvent = {
+          time: event.time,
+          value: midiPitchBend * Settings.getSetting("pitchbendrange")
+        };
+
+        MidiToNotes.pitchBendEvents.push(pitchEvent)
+      }
+    }
+  }
+
+  /***
+   * Finds the pitch adjust amount at a given time in the MIDI.
+   */
+  function getPitchBendAdjustmentAtTime(midiTime) {
+    let pitchEvent = {time:  -1, value: 0};
+    let nextPitchEvent = {time:  -1, value: 0};
+
+    for (let i = 0; i < MidiToNotes.pitchBendEvents.length; i++) {
+      if (MidiToNotes.pitchBendEvents[i].time <= midiTime) {
+        if (i + 1 >= MidiToNotes.pitchBendEvents.length) {
+          //Reached the last pitch bend event.
+          pitchEvent = MidiToNotes.pitchBendEvents[i];
+          return pitchEvent.value;
+        }
+        else if (MidiToNotes.pitchBendEvents[i + 1].time > midiTime) {
+          //Pitch bend is between the events before/after the current time.
+          pitchEvent = MidiToNotes.pitchBendEvents[i];
+          nextPitchEvent = MidiToNotes.pitchBendEvents[i + 1];
+
+          var timeDelta = nextPitchEvent.time - pitchEvent.time;
+          var pitchDelta = nextPitchEvent.value - pitchEvent.value;
+          
+          return pitchEvent.value + 
+            (((midiTime - pitchEvent.time) / timeDelta) * pitchDelta);
+        }
+      }
+    }
+
+    return 0.0;
+  }
+
+  /***
+   * Converts a given MIDI note pitch to TC pitch, and applies pitch bend.
+   * The note is clamped (and a warning shown) if it goes out of range.
+   */
+  function convertPitch(midiPitch, midiPitchBend, midiTime, timeDivision) {
+    const minTcPitch = -178.75;
+    const maxTcPitch = 178.75;
+    let tcPitch = (midiPitch - 60) * 13.75;
+    let tcPitchBend = midiPitchBend * 13.75;
+    let adjustedPitch = tcPitch + tcPitchBend;
+
+    if (adjustedPitch < minTcPitch || adjustedPitch > maxTcPitch) {
+      midiWarnings.add("Pitch bend adjustment clamped (out of range)", {
```

Comment:
I wasn't sure if the warnings were being used essentially as a user-visible logging that could be added to from anywhere? This was based on seeing a warning also being added in the main generateNotes function:

https://github.com/TC-Chart-Converter/TC-Chart-Converter.github.io/blob/88dcfc27e891a3a361dba5b2d29c1a3c179c0808/src/midiToNotes.js#L37

To integrate pitch bending into the warning function we'd need to deal with the interpolation between pitch bend events, the bend events cannot be clamped as that could impact other notes, so the note events would need to hold their post-pitch bend values (which can be verified and clamped)?

I think I prefer the alternative of moving the pitch warnings to the convert function, not sure if you agree but it makes sense to me having a note conversion function perform checks and add warnings if it made changes to the notes, rather than a generate warnings function make changes to the notes?

Reply:
If you think the note warnings should all be in the convert function, go for it

Reply:
Moved the unsnapped/out of range note checks into the convert function: 9650cf2038e55f97208f75e673eeca182e63c243
I combined the MIDI and post-pitch bend checks instead of checking the range twice, this now means a MIDI note could technically be out of range but put back into range by the pitch bend adjustment. GenerateWarnings now only contains the check for "meta" events.